### PR TITLE
Improve HB3A's IDF for calibration

### LINKED ISSUE #14087: HB3A's detector-sample distance may change every cycle
The distance between detector and sample of HB3A may change each cycle or even between 2 experiments.  Make its instrument definition file be more flexible such that take the distance as a parameter from log file.

### PR DESCRIPTION
**Pull request by wdzhou**
Description of work.

In this PR, a mistake on interpreting the rows and columns in XML file written by SPICE is corrected. The IDF of HB3A is modified such that the distance between the sample and detector plane can be set via log file.  The detector pixel that the neutron beam hits is moved from the geometry center of the detector to pixel (128, 115). 

**To test:**
1. Check the unit tests please.  All the new features are tested in the unit tests.
2. Load any HB3A data, for example, http://neutron.ornl.gov/user_data/hb3a/exp455/Datafiles/HB3A_exp455_scan0085_0018.xml.  If you view the detector from source in the instrument view, then you should see the data same to the image from that web page.
3. You may try to set an arbitrary value to sample-detector offset value to check how it is applied.

Fixes #14087.

[Release notes](/docs/source/release/)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by VickieLynch** [comment]
Works for my tests with HB3A data

:shipit:

**Comment by FedeMPouzols** [comment]
@wdzhou , @VickieLynch : could you write a summary of the testing done (in this case is fine as you explained that you had tested it with the relevant data), and tick the relevant items in the pull request description (for example "Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?" That makes the life of the gatekeepers easier so we can keep merging things in quickly.

File: Framework/DataHandling/inc/MantidDataHandling/LoadSpiceXML2DDet.h [hunk 1]
```diff
@@ -106,7 +111,28 @@ class DLLExport LoadSpiceXML2DDet : public API::Algorithm {
...
 
+  /// Set output workspace's X-axs as lab-frame Q space
   void setXtoLabQ(API::MatrixWorkspace_sptr dataws, const double &wavelength);
+
+  /// SPICE detector XML file
+  std::string m_detXMLFileName;
+  /// XML node name in detector counts file
+  std::string m_detXMLNodeName;
+  /// Pixel size at X direction
+  size_t m_numPixelX;
```
Comment (FedeMPouzols) [inline Framework/DataHandling/inc/MantidDataHandling/LoadSpiceXML2DDet.h]: I'd tend to prefer having these variables as local variables in the relevant methods (and passing them around if they are required in some other places). If there's no need to cache values having a more restricted scope usually helps making the code easier to read. Most algorithm classes do not need to have member variables (or they need very few).

File: Framework/DataHandling/inc/MantidDataHandling/LoadSpiceXML2DDet.h [hunk 2]
```diff
@@ -106,7 +111,28 @@ class DLLExport LoadSpiceXML2DDet : public API::Algorithm {
   /// Get wavelength from workspace
   bool getHB3AWavelength(API::MatrixWorkspace_sptr dataws, double &wavelength);
```
Comment (FedeMPouzols) [inline Framework/DataHandling/inc/MantidDataHandling/LoadSpiceXML2DDet.h]: This method could be `const`?

**Pull request merged**